tools/ir-generator/ir-generator.ypp

```diff
@@ -141,23 +140,38 @@ input
     ;
 
 element
-    : irclass           { global.push_back($1); }
-    | comment_block     { global.push_back($1); }
-    | comment_block irclass {
-            $2->comments.push_back($1);
-            global.push_back($2); }
-    | EMITBLOCK         { global.push_back(new EmitBlock(@1, $1.impl, $1.block)); }
+    : irclass {  
+        if (currCommentBlock) { 
+            $1->comments.push_back(currCommentBlock);
+            currCommentBlock = nullptr;
+        }
+        global.push_back($1); }
+    | COMMENTBLOCK  {
+        if (!currCommentBlock) {
+            currCommentBlock = new CommentBlock(@1, $1);
+        }
+        else {
+            currCommentBlock->append($1);
+            currCommentBlock->srcInfo += @1; }}
+    | EMITBLOCK { 
+        if (currCommentBlock) {
+            global.push_back(currCommentBlock);
+            currCommentBlock = nullptr;
+        }
```

Comment:
These 4 lines appear 3 times verbatim, so are a good candidate for factoring into a helper function.
(`pushCurrentComment` perhaps?)